lib/elixir/pages/anti-patterns/code-anti-patterns.md

```diff
@@ -51,11 +51,102 @@ TODO.
 
 ## Complex branching
 
-TODO.
+#### Problem
 
-## Complex else clauses in with
+When a function assumes the responsibility of handling multiple errors alone, it can increase its cyclomatic complexity (metric of control-flow) and become incomprehensible. This situation can configure a specific instance of "Long function", a traditional anti-pattern, but has implications of its own. Under these circumstances, this function could get very confusing, difficult to maintain and test, and therefore bug-proneness.
 
-TODO.
+#### Example
+
+An example of this anti-pattern is when a function uses the `case` control-flow structure or other similar constructs (for example, `cond`, or `receive`) to handle multiple variations of response types returned by the same API endpoint. This practice can make the function more complex, long, and difficult to understand, as shown next.
+
+```elixir
+def get_customer(customer_id) do
+  case get("/customers/#{customer_id}") do
+    {:ok, %Tesla.Env{status: 200, body: body}} -> {:ok, body}
+    {:ok, %Tesla.Env{body: body}} -> {:error, body}
+    {:error, _} = other -> other
+  end
+end
+```
+
+Although `get_customer/1` is not really long in this example, it could be. Thinking about this more complex scenario, where a large number of different responses can be provided to the same endpoint, is not a good idea to concentrate all on a single function. This is a risky scenario, where a little typo, or any problem introduced by the programmer in handling a response type, could eventually compromise the handling of all responses from the endpoint (if the function raises an exception, for example).
```

Comment:
I find this example doesn't illustrate the point really well, because in this precise case the "problematic" code looks clearer and more idiomatic to me 😅

How about making it look obviously problematic? We could use some ellipsis like this:

```elixir
    {:ok, %Tesla.Env{status: 200, body: body}} ->
      # complex processing done directly in the clause
      processed_response = ...
      {:ok, processed_response}
```

Reply:
Agreed. We need to say that each clause is made by several lines of code. The current example does not portray that. :)

Reply:
I have a feeling that employing ellipses in this instance would further abstract the example. What exactly constitutes "complex processing"? I would appreciate a more tangible code example that effectively illustrates the concept, but without making the documentation long and tiresome to read. :)

Reply:
How about something like

```elixir
body
|> Jason.decode!()
|> Map.fetch!("key")
|> String.trim()
```

Reply:
@fuelen we do cover more general anti-patterns, such as long parameters, so this one would fit. You are right, it applies to other cases, such as `if/else`, so perhaps that's worth a mention. It may also be worth mentioning that Elixir scopes make it _very_ easy to refactor. I can add some pseudo json-decoding to the function and refactor it from there.

Reply:
In my previous response, I wanted to mention that the function size is not so crucial for immutable languages, and moving code to a separate function doesn't always result in better code.



I suggest thinking about a different example. The current one exposes HTTP request details to the end user - in case of an error, Tesla can return `{:error, %Tesla.Error{}}` and the error contains HTTP request metadata. I think this is an antipattern as well. In the real world, people will use a middleware provided by Tesla to decode JSON.
Also, I think it is better to avoid the usage of 3rd-party libraries in these docs, as people will treat this as an officially-recommended library.


PS: I see there was a discussion about the previous version of the smell https://github.com/lucasvegi/Elixir-Code-Smells/issues/4 The original one had a smell mentioned here https://keathley.io/blog/good-and-bad-elixir.html in `Don’t pipe results into the following function` section

Reply:
Yup, I will refactor the Tesla part as well. 

>  I wanted to mention that the function size is not so crucial for immutable languages, 

That's why we removed the function size anti-pattern because it is hard to provide a specific rule for this. But, for complex branching, there is an easier argument to be made that the code is harder to grasp when the branches themselves are so far apart that you cannot see them at a glance.

So I will be going ahead with this one, including the changes above. :)